### PR TITLE
Aggregate rework

### PR DESCRIPTION
**Pull request by noxiouz**

### CODE REVIEW

File: aggregating/aggregating.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+				},
+				Config: i,
+				Data:   result,
+			}
+
+			logger.Debugf("%s data to send %v", task.Id, senderPayload)
+			payload, _ := common.Pack(senderPayload)
+			res, err := enqueue(app, payload)
+			if err != nil {
+				logger.Errf("%s unable to send to '%s' %s %s",
```
Comment (noxiouz) [inline aggregating/aggregating.go]: extra %s

File: aggregating/aggregating_test.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	"github.com/Combaine/Combaine/common/tasks"
+	"github.com/Sirupsen/logrus"
+)
+
+const (
+	cfgName  = "aggCore"
+	repoPath = "../tests/fixtures/configs"
+)
+
+var results chan []interface{} = make(chan []interface{})
```
Comment (noxiouz) [inline aggregating/aggregating_test.go]: var results = make(chan []interface{})

File: cmd/aggregating/main.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	raw := <-request.Read()
+	var task tasks.AggregationTask
+	err := common.Unpack(raw, &task)
+	if err != nil {
+		response.ErrorMsg(-100, err.Error())
+		return
+	}
+	err = aggregating.Aggregating(&task, cacher)
+	if err != nil {
+		response.ErrorMsg(-100, err.Error())
```
Comment (noxiouz) [inline cmd/aggregating/main.go]: return
adn drop else

File: aggregating/aggregating.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+				if !ok {
+					logger.Errf("%s 'perHost' support only bool value", task.Id)
+					continue
+				}
+				if !v {
+					continue
+				}
+
+				aggWg.Add(1)
+				go func(agg string, h string, c configs.PluginConfig, d *list,
```
Comment (noxiouz) [inline aggregating/aggregating.go]: why pointer to the list if list is []interface{} and it's not modified inside

File: aggregating/aggregating.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+					app servicecacher.Service, wg *sync.WaitGroup) {
+					defer wg.Done()
+
+					payload, _ := common.Pack(list{task.Id, c, *d})
```
Comment (noxiouz) [inline aggregating/aggregating.go]: []interface{}{....} is more understandable here in my point of view

File: aggregating/aggregating.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+				ch <- item{agg: agg, name: h, res: res}
+			}(name, subGroup, cfg, &subGroupParsingResults, app, &aggWg)
+		}
+		if len(aggParsingResults) == 0 {
+			logger.Infof("%s %s 'all' nothing aggregate", task.Id, name)
+			continue
+		}
+
+		aggWg.Add(1)
+		go func(agg string, h string, c configs.PluginConfig, d *list,
```
Comment (noxiouz) [inline aggregating/aggregating.go]: I see duplicated code here. May be it's better to do a free function?

File: aggregating/aggregating.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+			res, err := enqueue("aggregate_group", app, &payload, task.Id)
+			if err != nil {
+				logger.Errf("%s %s 'all' unable to aggregate %s", task.Id, agg, err)
+				return
+			}
+			ch <- item{agg: agg, name: h, res: res}
+		}(name, task.ParsingConfig.Metahost, cfg, &aggParsingResults, app, &aggWg)
+	}
+
+	go func() {
```
Comment (noxiouz) [inline aggregating/aggregating.go]: Why Wait is called in a separate goroutine? U can just Wait then close.

File: aggregating/aggregating.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	"sync"
+
+	"github.com/Combaine/Combaine/common"
+	"github.com/Combaine/Combaine/common/configs"
+	"github.com/Combaine/Combaine/common/logger"
+	"github.com/Combaine/Combaine/common/servicecacher"
+	"github.com/Combaine/Combaine/common/tasks"
+)
+
+type list []interface{}
```
Comment (noxiouz) [inline aggregating/aggregating.go]: space line)

File: aggregating/aggregating_test.go (hunk 2 above)
Comment (sakateka) [inline aggregating/aggregating_test.go]: ok

File: cmd/aggregating/main.go (hunk 3 above)
Comment (sakateka) [inline cmd/aggregating/main.go]: ok

File: aggregating/aggregating.go (hunk 5 above)
Comment (sakateka) [inline aggregating/aggregating.go]: ok)
I think the same

File: aggregating/aggregating.go (hunk 7 above)
Comment (sakateka) [inline aggregating/aggregating.go]: This allows the collection of aggregations as soon as they are ready.

File: aggregating/aggregating.go (hunk 6 above)
Comment (sakateka) [inline aggregating/aggregating.go]: ok

File: aggregating/aggregating.go (hunk 8 above)
Comment (sakateka) [inline aggregating/aggregating.go]: ok

File: aggregating/aggregating.go (hunk 4 above)
Comment (sakateka) [inline aggregating/aggregating.go]: ok

**Pull request merged**